R/filter_pd.R

```diff
@@ -0,0 +1,99 @@
+#'
+#' @examples
+#'adrs <- tibble::tribble(
+#'~STUDYID, ~USUBJID, ~PARAMCD, ~AVALC, ~ADT,
+#'"CDISCPILOT01", "01-701-1015", "OVR", "CR", "2016-01-25",
+#'"CDISCPILOT01", "01-701-1015", "OVR", "PD", "2016-01-25",
+#'"CDISCPILOT01", "01-701-1015", "BOR", "PD", "2016-02-21",
+#'"CDISCPILOT01", "01-701-1034", "OVR", "SD", "2015-07-12",
+#'"CDISCPILOT01", "01-701-1034", "OVR", "PD", "2016-04-25",
+#'"CDISCPILOT01", "01-701-1034", "OVR", "PD", "2016-06-25",
+#'"CDISCPILOT01", "01-701-1034", "BOR", "PD", "2016-05-25"
+#') %>% dplyr::mutate(
+#'  ADT = lubridate::as_date(ADT))
+#'
+#'adlb <- tibble::tribble(
+#'  ~STUDYID, ~USUBJID, ~ADT,
+#'  "CDISCPILOT01", "01-701-1015", "2014-02-13",
+#'  "CDISCPILOT01", "01-701-1015", "2015-05-21",
+#'  "CDISCPILOT01", "01-701-1015", "2015-12-21",
+#'  "CDISCPILOT01", "01-701-1015", "2016-07-08",
+#'  "CDISCPILOT01", "01-701-1015", "2016-11-25",
+#'  "CDISCPILOT01", "01-701-1015", "2017-02-21",
+#'  "CDISCPILOT01", "01-701-1034", "2015-04-03",
+#'  "CDISCPILOT01", "01-701-1034", "2016-01-05",
+#'  "CDISCPILOT01", "01-701-1034", "2016-04-02",
+#'  "CDISCPILOT01", "01-701-1034", "2016-04-25",
+#'  "CDISCPILOT01", "01-701-1034", "2016-06-25",
+#'  "CDISCPILOT01", "01-701-1034", "2016-05-25"
+#') %>% dplyr::mutate(
+#'  ADT = lubridate::as_date(ADT))
+#'
+#'  filter_pd(dataset = adlb,
+#'            filter_source = PARAMCD ==  "OVR" & AVALC == "PD",
+#'            source_pd = ADT,
+#'            source_dataset = adrs)
+filter_pd <- function(dataset,
+                      filter_source = PARAMCD == "OVR" & AVALC == "PD",
+                      source_pd = ADT,
+                      source_dataset = list(adrs = adrs)) {
+
+  assert_data_frame(dataset,
+                    required_vars = vars(STUDYID, USUBJID, ADT),
+                    optional = FALSE)
+  source_pd <- assert_symbol(enquo(source_pd))
+  assert_data_frame(source_dataset,
+                    required_vars = quo_c(source_pd,
+                                          vars(STUDYID, USUBJID, AVALC)),
+                                          optional = FALSE)
+  filter_source <- assert_filter_cond(enquo(filter_source),
+                                      optional = FALSE)
+
+  #filter source dataset and choose first PD
+  #based on condition specified by user
+  source_data <- filter_if(source_dataset,
+                           filter_source)
+  source_data <- filter_extreme(source_data,
+                                by_vars = vars(STUDYID, USUBJID),
+                                order =  vars(!!source_pd),
+                                mode = "FIRST")
+
+  #merge filtered source data and compare dates
+  #to keep only those up to source date
+  derive_vars_merged(dataset = dataset,
+                     dataset_add = source_data,
+                     by_vars = vars(STUDYID, USUBJID),
+                     order = vars(!!source_pd),
+                     new_vars = vars(temp_ADT = !!source_pd)) %>%
```

Comment:
It should be
```
derive_vars_merged(
  dataset = dataset,
  dataset_add = source_datasets[[source_pd$dataset_name]],
  filter_add = !!source_pd$filter,
  by_vars = vars(STUDYID, USUBJID),
  order = vars(!!source_pd$date),
  new_vars = vars(temp_ADT = !!source_pd$date),
  mode = "first") %>%
```
This would cover both cases that the first PD date is already derived and that it needs to be derived here.
```
source_pd = date_source(
  dataset_name = "adrs",
  date = ADT,
  filter = PARAMCD == "PD"
)
```
covers the first case and
```
source_pd = date_source(
  dataset_name = "adrs",
  date = ADT,
  filter = PARAMCD == "OVR" & ANL01FL == "Y" & AVALC == "PD"
)
```
the second one.

Reply:
so I am not sure why the design of this is to use date_source given the output of date_source is a list of dates that can be directly parsed into the function? From a user's perspective, a function that can be directly parsed is friendlier than having to understand two other functions before running the main function, no?